edgeless_orc/src/orchestrator.rs

```diff
@@ -725,6 +977,143 @@ impl Orchestrator {
                             }
                         }
                     }
+
+                    // Update the orchestration logic.
+                    orchestration_logic.update_nodes(&clients, &resource_providers);
+
+                    //
+                    // Make sure that all active logical functions are assigned
+                    // to one instance: for all the function instances that
+                    // were running in disconnected nodes, create new function
+                    // instances on other nodes, if possible and there were no
+                    // other running function instances.
+                    //
+
+                    // List of ext_fid that will have to be repatched
+                    // because of the allocation of new function instances
+                    // following node disconnection.
+                    let mut to_be_repatched = vec![]; // ext_fid
+
+                    // Function instances that have to be created to make up for
+                    // the loss of those assigned to disconnected nodes.
+                    // key:   ext_fid
+                    // value: function request
+                    let mut fun_to_be_created = std::collections::HashMap::new();
+
+                    // Resources that have to be created to make up for the
+                    // loss of those assigned to disconnected nodes.
+                    // key:   ext_fid
+                    // value: resource specs
+                    let mut res_to_be_created = std::collections::HashMap::new();
+
+                    // List of ext_fid that will have to be repatched.
+                    let mut active_instances_to_be_updated = vec![];
+
+                    // Find all the functions/resources affected.
+                    // Also attempt to start functions and resources that
+                    // are active but for which no active instance is present
+                    // (this happens because in the past a node with active
+                    // functions/resources has disappeared and it was not
+                    // possible to fix the situation immediately).
+                    for (origin_ext_fid, instance) in active_instances.iter() {
+                        match instance {
+                            ActiveInstance::Function(start_req, _reqs, instances) => {
+                                let num_disconnected = instances.iter().filter(|x| to_be_disconnected.contains(&x.node_id)).count();
+                                assert!(num_disconnected <= instances.len());
+                                if instances.is_empty() || num_disconnected > 0 {
+                                    to_be_repatched.push(origin_ext_fid.clone());
+                                    if instances.is_empty() || num_disconnected == instances.len() {
```

Comment:
This will lower the number of instances.
This might have cascading effects but can be dealt with later.

Reply:
The idea of the baseline orchestrator is that it only guarantees at _at least_ one instance is present for each logical function. More sophisticated orchestration policies can be done externally (or by a smarter version of the Orchestrator).